lib/SimpleSAML/Auth/Source.php

```diff
@@ -218,6 +218,116 @@ public function initLogin($return, $errorURL = null, array $params = [])
+        assert(is_callable($callback));
+        assert(is_array($params));
+
+        $session = Session::getSessionFromRequest();
+
+        $callbacks = $session->getData('\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks', $this->getAuthId());
+
+        if ($callbacks === null) {
+            $callbacks = [];
+        }
+        $callbacks[] = [$callback, $params];
+
+        $session->setData(
+            '\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks',
+            $this->getAuthId(),
+            $callbacks,
+            Session::DATA_TIMEOUT_SESSION_END
+        );
+    }
+
+
+    /**
+     * Check if authentication source consumer logout callbacks have been registered.
+     *
+     * @return bool
+     * @throws \Exception
+     */
+    public function hasAsConsumerLogoutCallbacks(): bool
+    {
+        $session = Session::getSessionFromRequest();
+        $callbacks = $session->getData('\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks', $this->getAuthId());
+        if ($callbacks === null) {
+            return false;
+        }
+
+        return !empty($callbacks);
+    }
+
+
+    /**
+     * Calls all registered logout callbacks for consumers of this source.
+     * This function might not return directly, so a $returnToUrl has to be specified.
+     *
+     * @param string $returnToUrl URL to return to if callback needs to leave this flow / redirect. Make sure to include the id parameter to recover state after processing.
+     * @throws \Exception
+     */
+    public function callAsConsumerLogoutCallbacks(string $returnToUrl)
+    {
+        assert(is_string($returnToUrl));
+
+        $state = [];
+        $state['asConsumerLogoutCallbacks:returnTo'] = $returnToUrl;
+        $state['asConsumerLogoutCallbacks:authId'] = $this->getAuthId();
+
+        self::handleAsConsumerLogoutCallbacks($state);
+    }
+
+
+    /**
+     * Handle authentication source consumer logout callbacks.
+     *
+     * @param $state
+     * @throws \Exception
+     */
+    public static function handleAsConsumerLogoutCallbacks($state)
+    {
+        assert(is_array($state));
+        assert(array_key_exists('asConsumerLogoutCallbacks:returnTo', $state));
+        assert(array_key_exists('asConsumerLogoutCallbacks:authId', $state));
+
+        $authId = $state['asConsumerLogoutCallbacks:authId'];
+        $session = \SimpleSAML\Session::getSessionFromRequest();
+        $callbacks = $session->getData('\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks', $authId);
+
+        if (!empty($callbacks)) {
+            // prepare updated returnToUrl in case callback redirects
+            $id = State::saveState($state, 'asConsumerLogoutCallbacks:resume');
+            $returnToUrl = Module::getModuleURL('core/source/as_logout_resume.php', ['id' => $id]);
+
+            foreach ($callbacks as $cb) { // FIXME: use loop variable or get rid of it!
+                $callback = array_shift($callbacks);
```

Comment:
Yes, please fix this

Reply:
Fixed in 82f1b492009472bdb6b866beada30efada687b4b